File: Python/2352. Equal Row and Column Pairs.py
from typing import List


# noinspection PyMethodMayBeStatic
class Solution:
    def equalPairs(self, grid: List[List[int]]) -> int:
        length = len(grid)
        rows = {}

        for i in range(length):
            cols = tuple(grid[i])
            rows[cols] = rows.get(cols, 0) + 1

        pairs = 0

        for i in range(length):
            cols = []
            for j in range(length):
                cols.append(grid[j][i])

            col = tuple(cols)
            if col in rows:
                pairs += rows[col]

        return pairs

File: JavaScript/2352. Equal Row and Column Pairs.js
// noinspection JSUnusedGlobalSymbols

/**
 * @param {number[][]} grid
 * @return {number}
 */
const equalPairs = (grid) => {
  const length = grid.length
  const rows = new Map()
  const cols = new Map()

  for (let i = 0; i < length; i++) {
    const col = []
    const row = []
    for (let j = 0; j < length; j++) {
      col.push(grid[i][j])
      row.push(grid[j][i])
    }
    const colKey = col.join('+')
    const rowKey = row.join('+')
    rows.set(colKey, (rows.get(colKey) || 0) + 1)
    cols.set(rowKey, (cols.get(rowKey) || 0) + 1)
  }

  let pairs = 0

  for (const [key, value] of cols.entries()) {
    if (rows.has(key)) {
      pairs += value * rows.get(key)
    }
  }

  return pairs
}
